src/main/java/touro/spf/InputManager.java

```diff
@@ -0,0 +1,92 @@
+package touro.spf;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.util.*;
+import java.util.regex.Pattern;
+
+/**
+ * Creates Node instances using data from input file. Passes
+ * the Nodes to graph.
+ */
+
+public class InputManager {
+
+    private final Scanner scanner;
+    private Graph graph;
+    private ArrayList<String> eachLine = new ArrayList<>();
+    private HashSet<Node> nodes = new HashSet<>();
+
+    public InputManager(String fileName, Graph graph) throws FileNotFoundException {
+        scanner = new Scanner(new FileInputStream(fileName));
+        this.graph = graph;
+        readFile();
+    }
+
+    private void readFile() {
+//        while (scanner.hasNextLine()) {
+        String next = scanner.nextLine();
+        while (!next.equals("0")) {
+            Node node, connection;
+
+            //an array of number chars and of length two
+            String[] strNumNode = next.split(" ");
+
+            //TODO: input validation. make sure only two numbers per line.
+
+            //check if node is in hash already
+            if (nodeIsInHashSet(strNumNode[0])) {
+                node = getNodeFromHashSet(strNumNode[0]);
+            } else {
+                //make new node, add to hash
+                node = new Node(strNumNode[0], new ArrayList<>());
+                nodes.add(node);
+            }
+
+            //check if connection is in hash already
+            if (nodeIsInHashSet(strNumNode[1])) {
+                connection = getNodeFromHashSet(strNumNode[1]);
+            } else {
+                connection = new Node(strNumNode[1], new ArrayList<>());
+                nodes.add(connection);
+            }
+
+            //check that node and connection are not connections of one
+            // another already
+            if(!node.isConnection(connection)){
+                node.addConnection(connection);
+            }
+            if(!connection.isConnection(node)){
+                connection.addConnection(node);
+            }
+            next = scanner.nextLine();
+        }
+//            eachLine.clear();
+//            nodes.clear();
+//        }
+        graph.buildGraph(nodes);
+    }
+
+    private Node getNodeFromHashSet(String nodeName) {
+
+        for (Node node : nodes) {
+            if (node.name.equals(nodeName)) {
+                return node;
+            }
+        } return null;
+    }
+
```

Comment:
I'm not totally following the context here, but it looks like getNodeFromHashSet gets the node with the name `nodeName`. 
I think there should be a way to do this by making a tempNode with nodeName, and then using `nodes.contains(tempNode)`, and you might have to override the `equals()` function (and possibly another hashing function as well?) inside of node. 
If that doesn't work out, this def works too, but it just kills the O(1) runtime by iterating through the nodes manually

Reply:
> I'm not totally following the context here, but it looks like getNodeFromHashSet gets the node with the name `nodeName`.
> I think there should be a way to do this by making a tempNode with nodeName, and then using `nodes.contains(tempNode)`, and you might have to override the `equals()` function (and possibly another hashing function as well?) inside of node.
> If that doesn't work out, this def works too, but it just kills the O(1) runtime by iterating through the nodes manually

Ah, yes! I knew there was a better way to do this! Thanks for the comment

Reply:
Ok, so I made a rather ignorant attempt at this. (I tried copying what Professor Schwimmer did in TouroSnake's Square class. The output I'm getting is way off... I have not taken the time to properly debug/test to see what it might be. I may just ask Professor Schwimmer how this whole contains(), hash(), equals() methods interact with each other and how I can implement it. You can check out my attempt [here](https://github.com/rachel613g/SPFGraphBuilder/tree/aBetterInputManager2).

Reply:
I made an attempt at this idea (as stated in another comment), but I actually had a question on how you are thinking it can be implemented in this scenario. Cause this method `getNodeFromHashSet` returns a Node. I tried to do what you were saying about assigning a tempNode, and I could think of a way to do that for the if statement that creates a new Node, but I couldn't see how that would help retrieve a node that already exists from the HashSet. It may be easier to undestand what I mean by checking out the code I attempted [here](https://github.com/rachel613g/SPFGraphBuilder/tree/aBetterInputManager2). You can see there how I make use of tempNode on lines 43 and 53 when making a new node, but how I still use the `getNodeFromHashSet` method on lines 40 and 50.